core: make messaging optional

kaspar030 opened the pull request:
In order to enhance RIOT's modularity, make messaging optional.

When not used but compiled in, messaging uses a whopping 24b of code on 32bit platforms, but doubles the size of each tcb (16b -> 40b).

Not sure the savings are worth the effort, but who knows...

(Added msg to DEFAULT_MODULES, so this will not affect anything but those brave souls who compile RIOT with ```DISABLE_MODULE += msg```).

OlegHahm commented:
Interesting

IldarValiev commented:
@kaspar030, may be it will be better to add new flag DISABLE_MESSAGING instead of MODULE_MSG - it makes more sense.

I think is important PR, because not in every case we need messaging, especially if RIOT controls simple sensor node.

OlegHahm commented:
Sorry, this got lost somehow. Could you rebase, @kaspar030?

kaspar030 commented:
- rebased

OlegHahm commented on tests/sizeof_tcb/Makefile:
```diff
@@ -1,4 +1,7 @@
 include ../Makefile.tests_common
 
+# optionally disable messaging
+#DISABLE_MODULE += msg
```
Wouldn't it make sense to comment this in here?

Maybe you could add this snippet also to `dist/Makefile`?

kaspar030 replied:
> Wouldn't it make sense to comment this in here?

hm?

OlegHahm replied:
I meant: why not indeed disabling the `msg` module here?

kaspar030 replied:
Ah. Nope, IMHO sizeof_tcb should show the tcb size in the general case. disabling messaging seems more on the special side.

OlegHahm replied:
Ah, yes, sorry, confused with the minimal example.

OlegHahm commented on Makefile.defaultmodules:
```diff
@@ -1,3 +1,3 @@
-DEFAULT_MODULE += board cpu core sys
+DEFAULT_MODULE += board cpu core sys msg
```
I would swap the order (makes more sense semantically and alphabetical).

OlegHahm commented:
I'm not sure, I see a concrete use case for that right now, but increasing modularity and making more features optional seems always a good idea if the overhead is manageable - which is the case for this change.

kaspar030 commented:
I've re-done the PR, it felt pointless to create a whole directory for msg, now I'm using ifdef within msg.c.

I've moved thread_print_msg_queue() to msg.c and renamed it to msg_queue_print(). @authmillenon, do you approve?

authmillenon commented:
I would prefer creating a new directory. Other then that `msg_queue_print()` is okay. :-)

authmillenon commented:
(especially since compilers will complain about the empty compile unit).

authmillenon commented:
((I saw enough of this type of "modulization" in lwip and emb6 that I know I don't want to have that in RIOT ;-)))

kaspar030 commented:
> I would prefer creating a new directory.

why?

> (especially since compilers will complain about the empty compile unit).
AFAIK, not if there's a single valid source line (e.g., headers).

authmillenon commented:
I gave already a bunch of reasons but here are more:

 * it's ugly
 * it gives a bad example and will enable future contributors to provide modules the same way
 * it's hard to assign binary data to modules (not in this example, but in lwIP - which uses *only* this kind of moduling - it definitely is a problem)

Compared to that:
 * moving it to a directory isn't hard ;-)

kaspar030 commented:
Martine, how is it possible that we end up having completely opposite opinions on so many matters? ;)

> I gave already a bunch of reasons but here are more:
No you didn't. You gave one, which I think is invalid.

>     it's ugly

Subjective. I don't agree. At least it is a lot less ugly than a subdirectory.

>     it gives a bad example and will enable future contributors to provide modules the same way

this is not the first module to do this.

>     it's hard to assign binary data to modules (not in this example, but in lwIP - which uses only this kind of moduling - it definitely is a problem)

Meaning - what?

> Compared to that:
> 
>     moving it to a directory isn't hard ;-)

It was in a subdirectory before the last commit(s). That looked ugly and completely unnecessary.

Mind you - the whole idea of disabling messaging is more an excercise of "because we can".
If it adds a pseudomodule and an #ifdef'd code block, fine. If it adds another directory (and thus becomes the only core .c file not in /core), it feels too intrusive to me for what it can do.

One-file modules in their own directory always seemed like a deficiency of the build system anyways.

OlegHahm commented:
Hm, difficult thing. On the one hand, having a subdirectory is kinda ugly, particular since it would be the only subfolder in `core`. On the other hand, this ifdef bracket around basically a whole file is kinda ugly, too. But finally, we cannot do this without some ifdefs anyway, so I'm willing to accept the current solution.

authmillenon commented:
Well if #4919 is imported and also added as a directory `msg` wouldn't be the only directory in core ;-). Also: I like directories they are not ugly :P.

authmillenon commented:
But fine. If I'm the only one that diggs subdirectories I'm fine with #ifdef blocks

authmillenon commented:
> this is not the first module to do this.

Which other pseudo-module does this? I see only dependency-collecting modules, modules that are only header files, common API modules, and "pseudo-submodules" (modules that add additional functionality to an existing module) in Makefile.pseudomodules. There is no module that is exactly like a normal module, but is a pseudo-module because the complete code for is in an `#ifdef`.

OlegHahm commented:
Needs a rebase.

haukepetersen commented:
I like the PR and I agree with @kaspar030, that I don't really like putting it into a separate folder. The only thing I think could make sense, is to rename the module to `core_msg`, to make clear that this module is part of core. @authmillenon can you live with this?

so 1st ACK

authmillenon commented:
Yeah.

kaspar030 commented:
- rebased, now waiting for #4103 as some lines are changed by both

kaspar030 commented:
Would someone give a second ACK?

kaspar030 commented:
green! :)

authmillenon commented on dist/Makefile:
```diff
@@ -36,6 +36,11 @@ QUIET ?= 1
+# If your application is very simple and doesn't use modules that use
+# messaging, it can be disabled to save some memory:
+
+#DISABLE_MODULE += msg
```
core_msg

authmillenon commented:
One minor doc thingy. Otherwise, fine... ACK

kaspar030 commented:
Thx! addressed & amended.
